OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs

```diff
@@ -77,26 +97,122 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
 
-			var actorCategorySelector = widget.Get<DropDownButtonWidget>("ACTOR_CATEGORY");
-			var filtersPanel = Ui.LoadWidget("ACTOR_CATEGORY_FILTER_PANEL", null, new WidgetArgs());
-			var categoryTemplate = filtersPanel.Get<CheckboxWidget>("CATEGORY_TEMPLATE");
 			var tileSetId = world.Map.Rules.TileSet.Id;
-			allCategories = mapRules.Actors.Where(a => !a.Value.Name.Contains('^')).Select(a => a.Value.TraitInfoOrDefault<EditorTilesetFilterInfo>())
-				.Where(i => i != null && i.Categories != null &&
-					!(i.ExcludeTilesets != null && i.ExcludeTilesets.Contains(tileSetId)) && !(i.RequireTilesets != null && !i.RequireTilesets.Contains(tileSetId)))
-				.SelectMany(i => i.Categories).Distinct().OrderBy(i => i).ToList();
-			selectedCategories.AddRange(allCategories);
+			var allActorsTemp = new List<ActorSelectorActor>();
+			foreach (var a in mapRules.Actors.Values)
+			{
+				// Partial templates are not allowed
+				if (a.Name.Contains('^'))
+					continue;
+
+				// Actor must have a preview associated with it
+				if (!a.HasTraitInfo<IRenderActorPreviewInfo>())
+					continue;
+
+				var editorData = a.TraitInfoOrDefault<EditorTilesetFilterInfo>();
+
+				// Actor must be included in at least one category
+				if (editorData == null || editorData.Categories == null)
+					continue;
+
+				// Excluded by tileset
+				if (editorData.ExcludeTilesets != null && editorData.ExcludeTilesets.Contains(tileSetId))
+					continue;
+
+				if (editorData.RequireTilesets != null && !editorData.RequireTilesets.Contains(tileSetId))
+					continue;
+
+				var tooltip = a.TraitInfos<EditorOnlyTooltipInfo>().FirstOrDefault(Exts.IsTraitEnabled) as TooltipInfoBase
+					?? a.TraitInfos<TooltipInfo>().FirstOrDefault(Exts.IsTraitEnabled);
+
+				var searchTerms = new List<string>() { a.Name };
+				searchTerms.AddRange(editorData.Categories);
+				if (tooltip != null)
+					searchTerms.Add(tooltip.Name);
+
+				var tooltipText = (tooltip == null ? "Type: " : tooltip.Name + "\nType: ") + a.Name;
+				allActorsTemp.Add(new ActorSelectorActor(a, editorData.Categories, searchTerms.ToArray(), tooltipText));
+ 			}
+
+			allActors = allActorsTemp.ToArray();
+
+			allCategories = allActors.SelectMany(ac => ac.Categories)
+				.Distinct()
+				.OrderBy(x => x)
+				.ToArray();
+
+			foreach (var c in allCategories)
+			{
+				selectedCategories.Add(c);
+				filteredCategories.Add(c);
+			}
+
+			var searchTextField = widget.GetOrNull<TextFieldWidget>("SEARCH_TEXTFIELD");
+			if (searchTextField != null)
+			{
+				searchTextField.OnTextEdited = () =>
+				{
+					searchFilter = searchTextField.Text.Trim();
+					filteredCategories.Clear();
+
+					if (!string.IsNullOrEmpty(searchFilter))
+						filteredCategories.AddRange(
+							allActors.Where(t => t.SearchTerms.Any(
+								s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
+							.SelectMany(t => t.Categories)
+							.Distinct()
+							.OrderBy(x => x));
```

Comment:
The `s => ...` is an inner expression of the `Where`, so is indented.  These other calls are all chained at the same level, so are indented at the same level.